### PR TITLE
Implement an unvectorized version of the COCORecall metric

### PR DESCRIPTION
**Pull request by LukeWood**
The PR is structured as follows:
- COCOBase counts bounding box matches, based on max_detections, iou_thresholds, an area_range, and category IDs
- COCOBase stores the results in three internal Tensors, TruePositives, FalsePositives, GroundTruthBoxes.
- COCORecall leverages these internally to compute the COCORecall metric

### CODE REVIEW

File: keras_cv/metrics/coco/base.py [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+
+        self.true_positives.assign_add(true_positives_update)
+        self.false_positives.assign_add(false_positives_update)
+        self.ground_truth_boxes.assign_add(ground_truth_boxes_update)
+
+    def _match_boxes(self, y_true, y_pred, threshold, ious):
+        n_true = tf.shape(y_true)[0]
+        n_pred = tf.shape(y_pred)[0]
+
+        gt_matches = tf.TensorArray(
```
[the remark this answers is not in the document]
Comment (bhack) [reply keras_cv/metrics/coco/base.py:145]: Are we touching  the interconversion limit here? https://www.tensorflow.org/xla/known_issues#tensorarray_tfxla_interconversion_is_not_supported

File: keras_cv/metrics/coco/base.py [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+        y_pred = util.sort_bboxes(y_pred, axis=bbox.CONFIDENCE)
+        true_positives_update = tf.zeros_like(self.true_positives)
+        false_positives_update = tf.zeros_like(self.false_positives)
+        ground_truth_boxes_update = tf.zeros_like(self.ground_truth_boxes)
+        
+        for img in tf.range(num_images):
+            sentinel_filtered_y_true = util.filter_out_sentinels(y_true[img])
+            sentinel_filtered_y_pred = util.filter_out_sentinels(y_pred[img])
+
+            area_filtered_y_true = util.filter_boxes_by_area_range(sentinel_filtered_y_true, self.area_range[0], self.area_range[1])
```
[the remark this answers is not in the document]
Comment (bhack) [reply keras_cv/metrics/coco/base.py:99]: Just a reminder I have pre-commit hooks waiting at https://github.com/keras-team/keras-cv/pull/14

File: keras_cv/metrics/coco/base.py [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+    def update_state(self, y_true, y_pred, sample_weight=None):
+        """
+        Args:
+            y_true: a bounding box Tensor in corners format.
+            y_pred: a bounding box Tensor in corners format.
+            sample_weight: Currently unsupported.
+        """
+        if sample_weight:
+            raise NotImplementedError(
+                "sample_weight is not yet supported in keras_cv COCO metrics."
```
[the remark this answers is not in the document]
Comment (LukeWood) [reply keras_cv/metrics/coco/base.py:82]: yeah, so I believe when we run our `assign_add` operation we'd just have to multiply by sample_weight[img_i]

Let's just wait until after vectorization, but this shouldn't be difficult.

File: keras_cv/metrics/coco/base.py [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+
+                ground_truth_boxes_update = tf.tensor_scatter_nd_add(
+                    ground_truth_boxes_update, [[k_i]], [tf.cast(tf.shape(ground_truths)[0], tf.float32)]
+                )
+
+        self.true_positives.assign_add(true_positives_update)
+        self.false_positives.assign_add(false_positives_update)
+        self.ground_truth_boxes.assign_add(ground_truth_boxes_update)
+
+    def _match_boxes(self, y_true, y_pred, threshold, ious):
```
[the remark this answers is not in the document]
Comment (LukeWood) [reply keras_cv/metrics/coco/base.py:156]: done

File: keras_cv/metrics/coco/base.py [hunk 5]
```diff
@@ -73,6 +73,7 @@ def reset_state(self):
         self.false_positives.assign(tf.zeros_like(self.false_positives))
         self.ground_truth_boxes.assign(tf.zeros_like(self.ground_truth_boxes))
 
+    @tf.function()
```
Comment (bhack) [inline keras_cv/metrics/coco/base.py:76]: Have you benchmarked the difference with:
https://www.tensorflow.org/api_docs/python/tf/function?version=nightly#using_type_annotations_to_improve_performance

Comment (LukeWood) [reply keras_cv/metrics/coco/base.py:76]: Not yet, I'm working on setting up thorough profiling now to figure out the best optimization strategy.

Comment (LukeWood) [reply keras_cv/metrics/coco/base.py:76]: I'll be sure to test out this compilation argument - thanks
